packages/react-router/src/router.ts

```diff
@@ -1797,91 +1778,146 @@ export class Router<
+                } catch (err) {
+                  beforeLoadPromise.reject(err)
                 }
 
-                const context = {
-                  ...parentContext,
-                  ...beforeLoadContext,
-                }
+                // Remove the beforeLoadPromise so future invocations will
+                // be fresh and not use the cached promise
+                updateMatch(matchId, (prev) => ({
+                  ...prev,
+                  beforeLoadPromise: undefined,
+                }))
 
-                matches[index] = match = {
-                  ...match,
-                  routeContext: replaceEqualDeep(
-                    match.routeContext,
-                    beforeLoadContext,
-                  ),
-                  context: replaceEqualDeep(match.context, context),
-                  abortController,
+                return await beforeLoadPromise
+              }
+
+              const handleBeforeLoadContext = (
+                beforeLoadContext: Record<string, any>,
+              ) => {
+                try {
+                  if (
+                    isRedirect(beforeLoadContext) ||
+                    isNotFound(beforeLoadContext)
+                  ) {
+                    handleSerialError(beforeLoadContext, 'BEFORE_LOAD')
+                  }
+
+                  updateMatch(matchId, (prev) => {
+                    const routeContext = {
+                      ...prev.routeContext,
+                      ...beforeLoadContext,
+                    }
+
+                    return {
+                      ...prev,
+                      routeContext: replaceEqualDeep(
+                        prev.routeContext,
+                        routeContext,
+                      ),
+                      context: replaceEqualDeep(prev.context, routeContext),
+                      abortController,
+                    }
+                  })
+                } catch (err) {
+                  handleSerialError(err, 'BEFORE_LOAD')
                 }
-                updateMatch(match.id, () => match)
+              }
+
+              if (shouldPending) {
+                // If we might show a pending component, we need to wait for the
+                // pending promise to resolve before we start showing that state
+                setTimeout(() => {
+                  try {
+                    checkLatest()
+                    // Update the match and prematurely resolve the loadMatches promise so that
+                    // the pending component can start rendering
+                    triggerOnReady()
+                  } catch {}
+                }, pendingMs)
+              }
+
+              const { paramsError, searchError } = getMatch()
+
+              if (paramsError) {
+                handleSerialError(paramsError, 'PARSE_PARAMS')
+              }
+
+              if (searchError) {
+                handleSerialError(searchError, 'VALIDATE_SEARCH')
+              }
+
+              try {
```

Comment:
I'd decorate this with a comment like "actually fire off the `beforeLoad` callback", purely for readability for others in the future.

Reply:
Resolved in https://github.com/TanStack/router/pull/1908